### DISCUSSION TITLE
DepthOfField pass, seems to bloom also? How to disable / control?

### DISCUSSION THREAD

**Discussion by supermoos** [pmndrs/postprocessing#294] [Q&A]
The Depth Of Field pass seems to also bloom / brighten the blurred areas, is there a way to control this bloom or disable it so the colors don not get brightened?

**Comment by vanruesc** [comment]
The `DepthOfFieldEffect` uses a special disc blur shader that produces bright circular highlights to simulate realistic bokeh visuals. The effect currently doesn't support alternative blur kernels, but it shouldn't be too hard to add such an option at a later time (after the planned convolution operator modularization). I'll add this to the roadmap for now.

**Reply by supermoos** [reply]
Sounds great! Could you point me to where I would need to look to alter this highlight to not highlight but instead create a bokeh effect like the one produced by `RealisticBokehEffect.js`? I've attached a few screenshots so you can better see what I'm trying to achieve. I really like the blur effects created by `RealisticBokehEffect.js` - however it seems to also blur / add a halo around the objects that should be in focus (i.e. not blurred) - any suggestions on how to remove that with `RealisticBokehEffect.js` - because then I could just use that one.

Correct focused objects with `DepthOfFieldEffect` but not a great look for the blurred objects:
<img width="2559" alt="Screenshot 2021-05-11 at 13 17 29" src="https://user-images.githubusercontent.com/2298345/117807360-d3327700-b25b-11eb-8bc8-54e0fd24f718.png">

Debug focused area shown with `RealisticBokehEffect.js`:
<img width="2560" alt="Screenshot 2021-05-11 at 13 18 41" src="https://user-images.githubusercontent.com/2298345/117807475-f5c49000-b25b-11eb-9b5b-6aa671031a00.png">

Halo effect around objects that shouldn't have any blur with `RealisticBokehEffect.js`:
<img width="2560" alt="Screenshot 2021-05-11 at 13 19 06" src="https://user-images.githubusercontent.com/2298345/117807556-0f65d780-b25c-11eb-8a72-d2d7deb91b16.png">

And here's a look reference for what I'm trying to make the blurred particles look like, this was made with `RealisticBokehEffect.js` with ManualDOF = false:
<img width="2560" alt="Screenshot 2021-05-11 at 13 25 13" src="https://user-images.githubusercontent.com/2298345/117807810-68357000-b25c-11eb-89cc-2199ae6785ad.png">

**Reply by vanruesc** [reply]
The `RealisticBokehEffect` was superceded by `DepthOfFieldEffect`. The blur kernels of both effects are fairly similar, but `DepthOfFieldEffect` uses a better algorithm. It can also perform the blurring at lower resolutions as opposed to the `RealisticBokehEffect` which always renders at full resolution.

What resolution are you using for the `DepthOfFieldEffect`? Have you tried using a `bokehScale` higher than 5?

**Reply by supermoos** [reply]
That all sounds great, it does also seem to perform a bit better and solve the issue of not actually blurring the objects I want to be in focus, but it seems to be rather limited (have unexposed uniforms?) compared to `RealisticBokehEffect` and the effects achievable with it, almost seems like the luminanceTreshold and luminanceGain values that I have turned to 0 in `RealisticBokehEffect` are missing as a tweakable option to bring down the highlight effect in `DepthOfFieldEffect`? 

I attached some screenshots here with `DepthOfFieldEffect` at higher bokehScale and different resolutions of 480 and 2560, sadly it doesn't come close to the look of `RealisticBokehEffect`

<img width="2560" alt="Screenshot 2021-05-11 at 16 47 07" src="https://user-images.githubusercontent.com/2298345/117836404-f0c20980-b278-11eb-9dee-b88aa774f7ef.png">

<img width="2560" alt="Screenshot 2021-05-11 at 16 47 15" src="https://user-images.githubusercontent.com/2298345/117836370-ea339200-b278-11eb-88d5-7a8ce8a734c7.png">

**Reply by vanruesc** [reply]
Internal uniforms shouldn't be regarded as part of the public API. Please use the constructor parameters and class members for configuration.

> seems like the luminanceTreshold and luminanceGain values [...] are missing

The settings `luminanceTreshold` and `luminanceGain` were not included in the new `DepthOfFieldEffect`, but there are plans to bring back fringe and gain at some point.

> I attached some screenshots here with DepthOfFieldEffect at higher bokehScale

That's too high, I would say. I think this first screenshot looked good:

<img width="300" src="https://user-images.githubusercontent.com/2298345/117807360-d3327700-b25b-11eb-8bc8-54e0fd24f718.png">

It probably just needed a slightly larger bokeh scale.

**Reply by supermoos** [reply]
> Internal uniforms shouldn't be regarded as part of the public API. Please use the constructor parameters and class members for configuration.
> 
Of course, let my rephrase, I was wondering if those uniforms could be exposed / reintroduced, or if they're hidden somewhere already and you could maybe point me to them so I could modify for my needs? :-)

> The settings `luminanceTreshold` and `luminanceGain` were not included in the new `DepthOfFieldEffect`, but there are plans to bring back fringe and gain at some point.
> 
Nice, would be great to get fringe back too!

> That's too high, I would say. I think this first screenshot looked good:
> 
> <img alt="" width="300" src="https://user-images.githubusercontent.com/2298345/117807360-d3327700-b25b-11eb-8bc8-54e0fd24f718.png">
Yes, except I'm trying to achieve the look of this screenshot for the particles that are blurred (if you notice, they have pronounced bokeh effect to them, where as the last screenshot you mentioned the particles more seem to disappear / brighten away. 
<img alt="" width="300" src="https://user-images.githubusercontent.com/2298345/117807810-68357000-b25c-11eb-89cc-2199ae6785ad.png">

And thanks for your time in looking at this! Really great to have a more standard "plug'n'play" to postprocessing with threejs.

**Reply by vanruesc** [reply]
> I was wondering if those uniforms could be exposed / reintroduced, or if they're hidden somewhere

I'm not aware of any hidden settings; every meaningful setting should already be accessible.

> they have pronounced bokeh effect to them, where as the last screenshot you mentioned the particles more seem to disappear / brighten away.

The screenshot of the `RealisticBokehEffect` looks like a uniform disc blur across the screen with no regard for depth. The `DepthOfFieldEffect` uses depth in a much more reliable manner, but depth effects heavily depend on a good depth range. What are your camera near/far and DoF focal length settings?

An example setup that is close to the desired result would be very useful for troubleshooting. At this time, I have no idea what settings might be misconfigured or if there's a real need for a new setting.

**Reply by supermoos** [reply]
Alright cool, I only suspected that perhaps the luminance value was somehow hardcoded because if I up the luminance treshold in `RealisticBokehEffect` it starts to resemble the look I get from `DepthOfFieldEffect`, so wanted to try and turn it down in `DepthOfFieldEffect`, but I gather there probably isn't. 

If you look to the far left top of the image there's a few particles that are actually in focus, but it seems `RealisticBokehEffect` just does a basic blend between a full blurred depth pass and the focused objects, and because the blur then exceeds the radius of the unblurred objects it bleeds trough. 

I can try and get an example setup up and going, completely understand the need! The camera settings are: 
```
        this._camera = new THREE.PerspectiveCamera(45, window.innerWidth / window.innerHeight, 1, 3000);
        this._camera.position.x = -1008;
        this._camera.position.y = -269.29;
        this._camera.position.z = -2220;
        this._camera.rotation.x = 3.02;
        this._camera.rotation.z = -0.42;
        this._camera.rotation.z = 3.09;
```
